Metrics

zhenlineo opened the pull request:
Exposing connection pool metrics for monitoring purpose.

Commit:
Rework on metrics.

Commit:
Simplified the new implementation

Commit:
Removed callback and added snapshot for better integration with quality tool

zhenlineo commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Driver statistics a.k.a. ConnectionPoolMetrics](https://trello.com/c/WHF1GlIj/352-driver-statistics-aka-connectionpoolmetrics)

ali-ince commented on driver/src/main/java/org/neo4j/driver/v1/Config.java:
```diff
@@ -111,11 +113,13 @@ private Config( ConfigBuilder builder)
     /**
-     * Logging provider
-     * @return the logging provider to use
+     * Logging metricsTracker
```
wrong find/replace? :)

ali-ince commented on driver/src/main/java/org/neo4j/driver/v1/ConnectionPoolMetrics.java:
```diff
@@ -0,0 +1,287 @@
+    PoolStatus poolStatus();
+
+    /**
+     * The amount of channels that are currently in-use (borrowed out of the pool).
```
maybe use `connections` instead of the driver internal concept `channels`?

ali-ince commented on driver/src/main/java/org/neo4j/driver/internal/metrics/ListenerEvent.java:
```diff
@@ -20,7 +20,7 @@
 {
-    void start();
-    long elapsed();
+    void start( long currentTime );
```
Wdyt if we leave the interface as earlier but pass the `Clock` instance to the `TimeRecorderListenerEvent`?

Commit:
Simplified `ListenerEvent` used in metrics

Commit:
Exposed metrics on Driver API.
Throws a client exception when accessing driver metrics if driver metrics is not enabled on driver Config.
